src/Configuration/HardwareManager/CMakeLists.txt

```diff
@@ -0,0 +1,89 @@
+# -*- mode:cmake -*-
+
+# Copyright 2023 The Foedag team
+
+# GPL License
+
+# Copyright (c) 2023 The Open-Source FPGA Foundation
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+cmake_minimum_required(VERSION 3.15)
+set(subsystem hardwaremanager)
+
+project(${subsystem} LANGUAGES CXX)
+
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+if (MSVC)
+else()
+  set(CMAKE_CXX_FLAGS_DEBUG
+  "${CMAKE_CXX_FLAGS_DEBUG} ${TCMALLOC_COMPILE_OPTIONS} -Werror -Wall -O0 -g ${MSYS_COMPILE_OPTIONS} ${MY_CXX_WARNING_FLAGS} ${MEM_SANITIZER_FLAGS}")
+  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -Werror")
+endif()
+
+#################################################################################################################
+#
+# Three elements will be built 
+#
+# 1. subsystem library (in this case hardwaremanager) - it contains all the core code to help configuration. It will be called by other elements
+#
+#################################################################################################################
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/../../../cmake/")
+find_package(USB1 REQUIRED)
+
+###################
+#
+# subsystem library
+#
+###################
+add_library(
+  ${subsystem} ${CFG_LIB_TYPE}
+  HardwareManager.cpp
+  OpenocdAdapter.cpp
+  OpenocdHelper.cpp
+)
+target_include_directories(${subsystem} PRIVATE ${LIBUSB_INCLUDE_DIR})
+target_link_libraries(${subsystem} PRIVATE ${LIBUSB_LIBRARIES})
+###################
+#
+# install 
+#   - for header file(s), only put those that we really need to export
+#
+###################
+set(
+  SRC_H_INSTALL_LIST
+)
+
+set(
+  SRC_LIB_INSTALL_LIST
+  ${subsystem}
+)
+
+foreach(file ${SRC_LIB_INSTALL_LIST})
+  install(
+    FILES ${CFG_BUILD_ROOT_DIR}/lib/$<TARGET_FILE_NAME:${file}>
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/foedag
+  )
+endforeach()
+
+foreach(file ${SRC_H_INSTALL_LIST})
+  install(
+    FILES ${PROJECT_SOURCE_DIR}/${file}
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/foedag/HardwareManager
+  )
+endforeach()
```

Comment:
add empty line

Reply:
<img width="950" alt="image" src="https://github.com/os-fpga/FOEDAG/assets/115125680/cfc656dd-ec63-48de-bc98-ee6a770e320e">
when running the formatting ./.github/bin/run-clang-format.sh, it does not allow empty line at end of file. It clean up.

```
guanyung@Lenovo-RS-GY:~/dev/FOEDAG$ ./.github/bin/run-clang-format.sh 
== There were changes running the formatter ==
diff --git a/src/Configuration/HardwareManager/CMakeLists.txt b/src/Configuration/HardwareManager/CMakeLists.txt
index eaa6f2ea..edcd9bde 100644
--- a/src/Configuration/HardwareManager/CMakeLists.txt
+++ b/src/Configuration/HardwareManager/CMakeLists.txt
@@ -86,4 +86,4 @@ foreach(file ${SRC_H_INSTALL_LIST})
     FILES ${PROJECT_SOURCE_DIR}/${file}
     DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/foedag/HardwareManager
   )
-endforeach()
\ No newline at end of file
+endforeach()
Submodule third_party/gtkwave contains modified content
To locally fix, run .github/bin/run-clang-format.sh then commit and push. 
```

Reply:
This is git diff warning telling us there is no newline at end of file. The clang formatter does allow this.